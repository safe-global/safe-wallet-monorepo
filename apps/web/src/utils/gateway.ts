import type { JsonRpcSigner } from 'ethers'
import { deleteTransaction } from '@safe-global/safe-gateway-typescript-sdk'
import { signTypedData } from './web3'

export const signTxServiceMessage = async (
  chainId: string,
  safeAddress: string,
  safeTxHash: string,
  signer: JsonRpcSigner,
): Promise<string> => {
  return await signTypedData(signer, {
    types: {
      DeleteRequest: [
        { name: 'safeTxHash', type: 'bytes32' },
        { name: 'totp', type: 'uint256' },
      ],
    },
    domain: {
      name: 'Safe Transaction Service',
      version: '1.0',
      chainId,
      verifyingContract: safeAddress,
    },
    message: {
      safeTxHash,
      totp: Math.floor(Date.now() / 3600e3),
    },
  })
}

export const deleteTx = async ({
  chainId,
  safeAddress,
  safeTxHash,
  signer,
}: {
  chainId: string
  safeAddress: string
  safeTxHash: string
  signer: JsonRpcSigner
}) => {
  const signature = await signTxServiceMessage(chainId, safeAddress, safeTxHash, signer)
  return await deleteTransaction(chainId, safeTxHash, signature)
}
