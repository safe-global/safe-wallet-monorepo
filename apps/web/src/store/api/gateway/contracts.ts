import { buildQueryFn } from '@/store/api/gateway'
import type { EndpointBuilder } from '@reduxjs/toolkit/query/react'
import type { fakeBaseQuery } from '@reduxjs/toolkit/query/react'
import { getContract as fetchContract } from '@safe-global/safe-gateway-typescript-sdk'
import type { Contract } from '@safe-global/safe-gateway-typescript-sdk/dist/types/contracts'

export const contractEndpoints = (
  builder: EndpointBuilder<ReturnType<typeof fakeBaseQuery<Error>>, 'OwnedSafes' | 'Submissions', 'gatewayApi'>,
) => ({
  getContract: builder.query<Contract, { chainId: string; contractAddress: string }>({
    queryFn({ chainId, contractAddress }) {
      return buildQueryFn(() => fetchContract(chainId, contractAddress))
    },
  }),
})
