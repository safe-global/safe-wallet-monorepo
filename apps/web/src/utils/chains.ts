import { AppRoutes } from '@/config/routes'
import type { ChainInfo } from '@safe-global/safe-gateway-typescript-sdk'
import { getExplorerLink } from '@safe-global/utils/utils/gateway'
import { FEATURES, hasFeature } from '@safe-global/utils/utils/chains'

export const FeatureRoutes = {
  [AppRoutes.apps.index]: FEATURES.SAFE_APPS,
  [AppRoutes.swap]: FEATURES.NATIVE_SWAPS,
  [AppRoutes.stake]: FEATURES.STAKING,
  [AppRoutes.balances.nfts]: FEATURES.ERC721,
  [AppRoutes.settings.notifications]: FEATURES.PUSH_NOTIFICATIONS,
  [AppRoutes.bridge]: FEATURES.BRIDGE,
  [AppRoutes.earn]: FEATURES.EARN,
}

export const getBlockExplorerLink = (
  chain: ChainInfo,
  address: string,
): { href: string; title: string } | undefined => {
  if (chain.blockExplorerUriTemplate) {
    return getExplorerLink(address, chain.blockExplorerUriTemplate)
  }
}

export const isRouteEnabled = (route: string, chain?: ChainInfo) => {
  if (!chain) return false
  const featureRoute = FeatureRoutes[route]
  return !featureRoute || hasFeature(chain, featureRoute)
}
