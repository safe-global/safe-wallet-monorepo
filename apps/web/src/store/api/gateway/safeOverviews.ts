import { type EndpointBuilder } from '@reduxjs/toolkit/query/react'

import type { SafeOverview } from '@safe-global/store/gateway/AUTO_GENERATED/safes'
import { sameAddress } from '@safe-global/utils/utils/addresses'
import type { RootState } from '../..'
import { selectCurrency } from '../../settingsSlice'
import { type SafeItem } from '@/features/myAccounts/hooks/useAllSafes'
import { asError } from '@safe-global/utils/services/exceptions/utils'
import { makeSafeTag } from '.'
import { additionalSafesRtkApi } from '@safe-global/store/gateway/safes'

type SafesInitiateThunk = ReturnType<typeof additionalSafesRtkApi.endpoints.safesGetOverviewForMany.initiate>
type SafesQueryActionResult = ReturnType<SafesInitiateThunk>
type SafesDispatch = (action: SafesInitiateThunk) => SafesQueryActionResult

type SafeOverviewQueueItem = {
  safeAddress: string
  walletAddress?: string
  chainId: string
  currency: string
  dispatch: SafesDispatch
  callback: (result: { data: SafeOverview | undefined; error?: never } | { data?: never; error: string }) => void
}

const _BATCH_SIZE = 10
const _FETCH_TIMEOUT = 300

class SafeOverviewFetcher {
  private requestQueue: SafeOverviewQueueItem[] = []

  private fetchTimeout: NodeJS.Timeout | null = null

  private async fetchSafeOverviews({
    safeIds,
    walletAddress,
    currency,
    dispatch,
  }: {
    safeIds: `${number}:0x${string}`[]
    walletAddress?: string
    currency: string
    dispatch: SafesDispatch
  }) {
    const queryThunk = additionalSafesRtkApi.endpoints.safesGetOverviewForMany.initiate({
      safes: safeIds,
      currency,
      walletAddress,
      trusted: false,
      excludeSpam: true,
    })
    const queryAction: SafesQueryActionResult = dispatch(queryThunk)

    try {
      return await queryAction.unwrap()
    } finally {
      queryAction.unsubscribe()
    }
  }

  private async processQueuedItems() {
    // Dequeue the first BATCH_SIZE items
    const nextBatch = this.requestQueue.slice(0, _BATCH_SIZE)
    this.requestQueue = this.requestQueue.slice(_BATCH_SIZE)

    let overviews: SafeOverview[]
    try {
      this.fetchTimeout && clearTimeout(this.fetchTimeout)
      this.fetchTimeout = null

      if (nextBatch.length === 0) {
        // Nothing to process
        return
      }

      const safeIds = nextBatch.map((request) => makeSafeTag(request.chainId, request.safeAddress))
      const { walletAddress, currency, dispatch } = nextBatch[0]
      overviews = await this.fetchSafeOverviews({ safeIds, currency, walletAddress, dispatch })
    } catch (err) {
      // Overviews could not be fetched
      nextBatch.forEach((item) => item.callback({ error: 'Could not fetch Safe overview' }))
      return
    }

    nextBatch.forEach((item) => {
      const overview = overviews.find(
        (entry) => sameAddress(entry.address.value, item.safeAddress) && entry.chainId === item.chainId,
      )

      item.callback({ data: overview })
    })
  }

  private enqueueRequest(item: SafeOverviewQueueItem) {
    this.requestQueue.push(item)

    if (this.requestQueue.length >= _BATCH_SIZE) {
      this.processQueuedItems()
    }

    // If no timer is running start a timer
    if (this.fetchTimeout === null) {
      this.fetchTimeout = setTimeout(() => {
        this.processQueuedItems()
      }, _FETCH_TIMEOUT)
    }
  }

  async getOverview(item: Omit<SafeOverviewQueueItem, 'callback'>) {
    return new Promise<SafeOverview | undefined>((resolve, reject) => {
      this.enqueueRequest({
        ...item,
        callback: (result) => {
          if ('data' in result) {
            resolve(result.data)
          }
          reject(result.error)
        },
      })
    })
  }
}

const batchedFetcher = new SafeOverviewFetcher()

type MultiOverviewQueryParams = {
  currency: string
  walletAddress?: string
  safes: SafeItem[]
}

export const safeOverviewEndpoints = (builder: EndpointBuilder<any, 'OwnedSafes' | 'Submissions', 'gatewayApi'>) => ({
  getSafeOverview: builder.query<SafeOverview | null, { safeAddress: string; walletAddress?: string; chainId: string }>(
    {
      async queryFn({ safeAddress, walletAddress, chainId }, { getState, dispatch }) {
        const state = getState() as RootState
        const currency = selectCurrency(state)
        const dispatchSafeOverview: SafesDispatch = (action) => dispatch(action)

        if (!safeAddress) {
          return { data: null }
        }

        try {
          const safeOverview = await batchedFetcher.getOverview({
            chainId,
            currency,
            walletAddress,
            safeAddress,
            dispatch: dispatchSafeOverview,
          })
          return { data: safeOverview ?? null }
        } catch (error) {
          return { error: { status: 'CUSTOM_ERROR', error: asError(error).message } }
        }
      },
    },
  ),
  getMultipleSafeOverviews: builder.query<SafeOverview[], MultiOverviewQueryParams>({
    async queryFn(params, { dispatch }) {
      const { safes, walletAddress, currency } = params
      const dispatchSafeOverview: SafesDispatch = (action) => dispatch(action)

      try {
        const promisedSafeOverviews = safes.map((safe) =>
          batchedFetcher.getOverview({
            chainId: safe.chainId,
            safeAddress: safe.address,
            currency,
            walletAddress,
            dispatch: dispatchSafeOverview,
          }),
        )
        const safeOverviews = await Promise.all(promisedSafeOverviews)
        return { data: safeOverviews.filter(Boolean) as SafeOverview[] }
      } catch (error) {
        return { error: { status: 'CUSTOM_ERROR', error: (error as Error).message } }
      }
    },
  }),
})
