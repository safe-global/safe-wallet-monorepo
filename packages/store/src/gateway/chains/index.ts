import { type Chain as ChainInfo } from '../AUTO_GENERATED/chains'
import { createEntityAdapter, EntityState } from '@reduxjs/toolkit'
import { cgwClient, getBaseUrl } from '../cgwClient'
import { QueryReturnValue, FetchBaseQueryError, FetchBaseQueryMeta } from '@reduxjs/toolkit/dist/query'

export const chainsAdapter = createEntityAdapter<ChainInfo, string>({ selectId: (chain: ChainInfo) => chain.chainId })
export const initialState = chainsAdapter.getInitialState()

const getChainsConfigs = async (
  url = `${getBaseUrl()}/v1/chains`,
  results: ChainInfo[] = [],
): Promise<EntityState<ChainInfo, string>> => {
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`)
  }
  const data = await response.json()

  const nextResults = [...results, ...data.results]

  if (data.next) {
    return getChainsConfigs(data.next, nextResults)
  }

  return chainsAdapter.setAll(initialState, nextResults)
}

const getChains = async (): Promise<
  QueryReturnValue<EntityState<ChainInfo, string>, FetchBaseQueryError, FetchBaseQueryMeta>
> => {
  try {
    const data = await getChainsConfigs()
    return { data }
  } catch (error) {
    return { error: error as FetchBaseQueryError }
  }
}

export const apiSliceWithChainsConfig = cgwClient.injectEndpoints({
  endpoints: (builder) => ({
    getChainsConfig: builder.query<EntityState<ChainInfo, string>, void>({
      queryFn: async () => {
        return getChains()
      },
    }),
  }),
  overrideExisting: true,
})
