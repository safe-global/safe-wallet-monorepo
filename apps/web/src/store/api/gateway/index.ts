import { proposerEndpoints } from '@/store/api/gateway/proposers'
import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react'

import {
  type AllOwnedSafes,
  getAllOwnedSafes,
  getTransactionDetails,
  type TransactionDetails,
} from '@safe-global/safe-gateway-typescript-sdk'
import { asError } from '@/services/exceptions/utils'
import { safeOverviewEndpoints } from './safeOverviews'
import { createSubmission, getSafe, getSafesByOwner, getSubmission } from '@safe-global/safe-client-gateway-sdk'

export async function buildQueryFn<T>(fn: () => Promise<T>) {
  try {
    return { data: await fn() }
  } catch (error) {
    return { error: asError(error) }
  }
}

export function makeSafeTag(chainId: string, address: string): `${number}:0x${string}` {
  return `${chainId}:${address}` as `${number}:0x${string}`
}

export const gatewayApi = createApi({
  reducerPath: 'gatewayApi',
  baseQuery: fakeBaseQuery<Error>(),
  tagTypes: ['OwnedSafes', 'Submissions'],
  endpoints: (builder) => ({
    getTransactionDetails: builder.query<TransactionDetails, { chainId: string; txId: string }>({
      queryFn({ chainId, txId }) {
        return buildQueryFn(() => getTransactionDetails(chainId, txId))
      },
    }),
    getMultipleTransactionDetails: builder.query<TransactionDetails[], { chainId: string; txIds: string[] }>({
      queryFn({ chainId, txIds }) {
        return buildQueryFn(() => Promise.all(txIds.map((txId) => getTransactionDetails(chainId, txId))))
      },
    }),
    getSafe: builder.query<getSafe, { chainId: string; safeAddress: string }>({
      queryFn({ chainId, safeAddress }) {
        return buildQueryFn(() => getSafe({ params: { path: { chainId, safeAddress } } }))
      },
    }),
    getAllOwnedSafes: builder.query<AllOwnedSafes, { walletAddress: string }>({
      queryFn({ walletAddress }) {
        return buildQueryFn(() => getAllOwnedSafes(walletAddress))
      },
    }),
    getOwnedSafes: builder.query<getSafesByOwner, { chainId: string; ownerAddress: string }>({
      queryFn({ chainId, ownerAddress }) {
        return buildQueryFn(() => getSafesByOwner({ params: { path: { chainId, ownerAddress } } }))
      },
      providesTags: (_res, _err, { chainId, ownerAddress }) => {
        return [{ type: 'OwnedSafes', id: makeSafeTag(chainId, ownerAddress) }]
      },
    }),
    getSubmission: builder.query<
      getSubmission,
      { outreachId: number; chainId: string; safeAddress: string; signerAddress: string }
    >({
      queryFn({ outreachId, chainId, safeAddress, signerAddress }) {
        return buildQueryFn(() =>
          getSubmission({ params: { path: { outreachId, chainId, safeAddress, signerAddress } } }),
        )
      },
      providesTags: ['Submissions'],
    }),
    createSubmission: builder.mutation<
      createSubmission,
      { outreachId: number; chainId: string; safeAddress: string; signerAddress: string }
    >({
      queryFn({ outreachId, chainId, safeAddress, signerAddress }) {
        return buildQueryFn(() =>
          createSubmission({
            params: {
              path: { outreachId, chainId, safeAddress, signerAddress },
            },
            body: { completed: true },
          }),
        )
      },
      invalidatesTags: ['Submissions'],
    }),
    ...proposerEndpoints(builder),
    ...safeOverviewEndpoints(builder),
  }),
})

export const {
  useGetTransactionDetailsQuery,
  useGetMultipleTransactionDetailsQuery,
  useLazyGetTransactionDetailsQuery,
  useGetProposersQuery,
  useDeleteProposerMutation,
  useAddProposerMutation,
  useGetSubmissionQuery,
  useCreateSubmissionMutation,
  useGetSafeQuery,
  useGetSafeOverviewQuery,
  useGetMultipleSafeOverviewsQuery,
  useGetAllOwnedSafesQuery,
  useGetOwnedSafesQuery,
} = gatewayApi
