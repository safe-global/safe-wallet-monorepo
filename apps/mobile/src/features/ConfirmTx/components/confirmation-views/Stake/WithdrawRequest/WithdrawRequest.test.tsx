import React from 'react'
import { renderWithStore, createTestStore } from '@/src/tests/test-utils'
import { StakingWithdrawRequest } from './WithdrawRequest'
import {
  NativeStakingValidatorsExitTransactionInfo,
  MultisigExecutionDetails,
  Operation,
} from '@safe-global/store/gateway/AUTO_GENERATED/transactions'
import { apiSliceWithChainsConfig } from '@safe-global/store/gateway/chains'

const mockWithdrawRequestTxInfo: NativeStakingValidatorsExitTransactionInfo = {
  type: 'NativeStakingValidatorsExit',
  humanDescription: 'Request withdrawal of staked tokens',
  status: 'ACTIVE',
  estimatedExitTime: 30 * 86400000, // 30 days in milliseconds
  estimatedWithdrawalTime: 2 * 86400000, // 2 days in milliseconds
  value: '32000000000000000000', // 32 ETH in wei
  numValidators: 1,
  tokenInfo: {
    address: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
    decimals: 18,
    logoUri: 'https://safe-transaction-assets.safe.global/chains/1/chain_logo.png',
    name: 'Ethereum',
    symbol: 'ETH',
    trusted: true,
  },
  validators: ['0x123...abc'],
}

const mockExecutionInfo: MultisigExecutionDetails = {
  type: 'MULTISIG',
  submittedAt: 1234567890,
  nonce: 1,
  safeTxGas: '0',
  baseGas: '0',
  gasPrice: '0',
  gasToken: '0x0000000000000000000000000000000000000000',
  refundReceiver: {
    value: '0x0000000000000000000000000000000000000000',
  },
  safeTxHash: '0x123',
  signers: [],
  confirmationsRequired: 2,
  confirmations: [],
  rejectors: [],
  trusted: true,
}

const mockTxData = {
  to: {
    value: '0x1234567890123456789012345678901234567890',
    name: 'Staking Contract',
    logoUri: null,
  },
  operation: 0 as Operation,
}

const mockProps = {
  txInfo: mockWithdrawRequestTxInfo,
  executionInfo: mockExecutionInfo,
  txId: 'test-tx-id',
  txData: mockTxData,
}

describe('StakingWithdrawRequest', () => {
  let store: ReturnType<typeof createTestStore>

  beforeEach(async () => {
    store = createTestStore({
      activeSafe: {
        address: '0x1234567890123456789012345678901234567890',
        chainId: '1',
      },
    })

    await store.dispatch(apiSliceWithChainsConfig.endpoints.getChainsConfig.initiate(undefined))
  })

  it('renders correctly with withdraw request information', () => {
    const { getByText, getAllByText } = renderWithStore(<StakingWithdrawRequest {...mockProps} />, store)

    expect(getAllByText(/32.*ETH/)).toHaveLength(2) // TokenAmount in header and receive row
    expect(getByText('Exit')).toBeTruthy() // Exit label
    expect(getByText('Validator 1')).toBeTruthy() // Validator link
    expect(getAllByText('Receive')).toHaveLength(2) // Receive label (header and table)
    expect(getByText('Withdraw in')).toBeTruthy() // Withdraw in label
    expect(getByText(/Up to/)).toBeTruthy() // Withdraw timing
    expect(getByText(/withdrawal request/)).toBeTruthy() // Warning message
    expect(getByText(/Dedicated Staking for ETH/)).toBeTruthy() // Description
  })

  it('renders multiple validators correctly', () => {
    const multiValidatorProps = {
      ...mockProps,
      txInfo: {
        ...mockWithdrawRequestTxInfo,
        numValidators: 3,
        validators: ['0x123...abc', '0x456...def', '0x789...ghi'],
      },
    }

    const { getByText } = renderWithStore(<StakingWithdrawRequest {...multiValidatorProps} />, store)

    expect(getByText('Exit')).toBeTruthy()
    expect(getByText('Validator 1')).toBeTruthy()
    expect(getByText('Validator 2')).toBeTruthy()
    expect(getByText('Validator 3')).toBeTruthy()
  })

  it('matches snapshot', () => {
    const component = renderWithStore(<StakingWithdrawRequest {...mockProps} />, store)
    expect(component).toMatchSnapshot()
  })
})
