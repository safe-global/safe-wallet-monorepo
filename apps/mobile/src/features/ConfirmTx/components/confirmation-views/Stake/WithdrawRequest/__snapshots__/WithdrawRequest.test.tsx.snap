// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StakingWithdrawRequest matches snapshot 1`] = `
<View>
  <View
    style={
      {
        "flexDirection": "column",
        "gap": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "column",
          "gap": 8,
          "marginTop": 16,
          "position": "relative",
        }
      }
    >
      <View
        style={
          {
            "width": 40,
          }
        }
      >
        <View
          style={
            {
              "position": "absolute",
              "right": -10,
              "top": -10,
              "zIndex": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#DDDEE0",
                "borderBottomColor": "#EEEFF0",
                "borderBottomLeftRadius": 100000,
                "borderBottomRightRadius": 100000,
                "borderBottomWidth": 1,
                "borderLeftColor": "#EEEFF0",
                "borderLeftWidth": 1,
                "borderRightColor": "#EEEFF0",
                "borderRightWidth": 1,
                "borderStyle": "solid",
                "borderTopColor": "#EEEFF0",
                "borderTopLeftRadius": 100000,
                "borderTopRightRadius": 100000,
                "borderTopWidth": 1,
                "flexDirection": "column",
                "height": 24,
                "justifyContent": "center",
                "maxHeight": 24,
                "maxWidth": 24,
                "minHeight": 24,
                "minWidth": 24,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
                "width": 24,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#A1A3A7",
                    "fontSize": 12,
                  },
                  undefined,
                  {
                    "fontFamily": "SafeIcons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              î¥·
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "#EEEFF0",
              "borderBottomLeftRadius": "50%",
              "borderBottomRightRadius": "50%",
              "borderTopLeftRadius": "50%",
              "borderTopRightRadius": "50%",
              "height": 40,
              "width": 40,
            }
          }
        >
          <ViewManagerAdapter_ExpoImage
            borderRadius={50}
            containerViewRef={"[React.ref]"}
            contentFit="cover"
            contentPosition={
              {
                "left": "50%",
                "top": "50%",
              }
            }
            flex={1}
            nativeViewRef={"[React.ref]"}
            onError={[Function]}
            onLoad={[Function]}
            onLoadStart={[Function]}
            onProgress={[Function]}
            placeholder={[]}
            source={
              [
                {
                  "uri": "https://safe-transaction-assets.safe.global/chains/1/chain_logo.png",
                },
              ]
            }
            style={
              {
                "borderRadius": 50,
                "flex": 1,
              }
            }
            testID="logo-image"
            transition={null}
          />
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "gap": 8,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "gap": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "#121312",
                "fontFamily": "DM Sans",
              }
            }
            suppressHighlighting={true}
          >
            Receive
          </Text>
          <Text
            style={
              {
                "color": "#121312",
                "fontFamily": "DMSans-Bold",
              }
            }
            suppressHighlighting={true}
          >
            32
             
            ETH
          </Text>
        </View>
        <Text
          style={
            {
              "color": "#A1A3A7",
              "fontFamily": "DM Sans",
              "fontSize": 12,
              "lineHeight": 16,
            }
          }
          suppressHighlighting={true}
        >
          15 Jan 1970
          , 
          06:56 AM
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 7,
          "borderBottomRightRadius": 7,
          "borderTopLeftRadius": 7,
          "borderTopRightRadius": 7,
          "flexDirection": "column",
          "gap": 20,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            {
              "color": "#A1A3A7",
              "flex": 1,
              "fontFamily": "DM Sans",
              "fontSize": 14,
            }
          }
          suppressHighlighting={true}
        >
          Exit
        </Text>
        <Text
          style={
            {
              "color": "#121312",
              "flex": 2,
              "fontFamily": "DM Sans",
              "fontSize": 14,
              "textAlign": "right",
            }
          }
          suppressHighlighting={true}
        >
          1 Validator
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            {
              "color": "#A1A3A7",
              "flex": 1,
              "fontFamily": "DM Sans",
              "fontSize": 14,
            }
          }
          suppressHighlighting={true}
        >
          Receive
        </Text>
        <Text
          style={
            {
              "color": "#121312",
              "fontFamily": "DMSans-SemiBold",
            }
          }
          suppressHighlighting={true}
        >
          32
           
          ETH
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            {
              "color": "#A1A3A7",
              "flex": 1,
              "fontFamily": "DM Sans",
              "fontSize": 14,
            }
          }
          suppressHighlighting={true}
        >
          Withdraw in
        </Text>
        <Text
          style={
            {
              "color": "#121312",
              "flex": 2,
              "fontFamily": "DM Sans",
              "fontSize": 14,
              "textAlign": "right",
            }
          }
          suppressHighlighting={true}
        >
          Up to 1 day
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "height": 40,
          }
        }
      >
        <View
          focusVisibleStyle={{}}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          role="button"
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#DCDEE0",
              "borderBottomColor": "transparent",
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderBottomWidth": 0,
              "borderLeftColor": "transparent",
              "borderLeftWidth": 0,
              "borderRightColor": "transparent",
              "borderRightWidth": 0,
              "borderStyle": "solid",
              "borderTopColor": "transparent",
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
              "borderTopWidth": 0,
              "bottom": 0,
              "flexDirection": "row",
              "flexWrap": "nowrap",
              "fontWeight": "700",
              "height": 36,
              "justifyContent": "center",
              "left": 0,
              "paddingLeft": 8,
              "paddingRight": 8,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <Text
            lineBreakMode="clip"
            numberOfLines={1}
            style={
              {
                "color": "#121312",
                "flexGrow": 0,
                "flexShrink": 1,
                "fontFamily": "DMSans-Bold",
                "fontSize": 14,
                "letterSpacing": 0,
                "lineHeight": 15.400000000000002,
                "userSelect": "none",
              }
            }
            suppressHighlighting={true}
          >
            Transaction details
          </Text>
        </View>
      </View>
      <Text
        style={
          {
            "color": "#A1A3A7",
            "fontFamily": "DM Sans",
            "fontSize": 13,
          }
        }
        suppressHighlighting={true}
      >
        The selected amount and any rewards will be withdrawn from Dedicated Staking for ETH after the validator exit.
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "column",
          "gap": 12,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#FFECC2",
                "borderBottomLeftRadius": 5,
                "borderBottomRightRadius": 5,
                "borderTopLeftRadius": 5,
                "borderTopRightRadius": 5,
                "flexDirection": "row",
                "gap": 12,
                "justifyContent": "center",
                "paddingBottom": 8,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 8,
                "width": "100%",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FF8C00",
                  "borderBottomLeftRadius": 34,
                  "borderBottomRightRadius": 34,
                  "borderTopLeftRadius": 34,
                  "borderTopRightRadius": 34,
                  "display": "flex",
                  "justifyContent": "center",
                  "paddingBottom": 4,
                  "paddingLeft": 4,
                  "paddingRight": 4,
                  "paddingTop": 4,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#FFFFFF",
                      "fontSize": 16,
                    },
                    undefined,
                    {
                      "fontFamily": "SafeIcons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="warning-icon"
              >
                î¤
              </Text>
            </View>
            <View
              style={
                {
                  "flexShrink": 1,
                  "gap": 4,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#6C2D19",
                    "fontFamily": "DMSans-SemiBold",
                    "fontSize": 14,
                  }
                }
                suppressHighlighting={true}
              >
                This transaction is a withdrawal request. After it's executed, you'll need to complete a separate withdrawal transaction.
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
